lib/ace/transport_app.rb

```diff
@@ -215,37 +217,32 @@ def self.trusted_facts(certname)
       # workflow, errors within the Configuer and not as pleasant - can have
       # _some_ control over them, especially around status codes from
       # the /v4/catalog endpoint
-      @plugins.with_synced_libdir(environment, certname) do
-        ACE::TransportApp.init_puppet_target(certname, body['target']['remote-transport'], body['target'])
-        configurer = ACE::Configurer.new(body['compiler']['transaction_uuid'], body['compiler']['job_id'])
-        configurer.run(transport_name: certname,
-                       environment: environment,
-                       network_device: true,
-                       pluginsync: false,
-                       trusted_facts: ACE::TransportApp.trusted_facts(certname))
-      end
-
-      # simulate expected error cases
-      if body['compiler']['certname'] == 'fail.example.net'
-        [200, { _error: {
-          msg: 'catalog compile failed',
-          kind: 'puppetlabs/ace/compile_failed',
-          details: 'upstream api errors go here'
-        } }.to_json]
-      elsif body['compiler']['certname'] == 'credentials.example.net'
-        [200, { _error: {
-          msg: 'target specification invalid',
-          kind: 'puppetlabs/ace/target_spec',
-          details: 'upstream api errors go here'
-        } }.to_json]
-      elsif body['compiler']['certname'] == 'reports.example.net'
-        [200, { _error: {
-          msg: 'report submission failed',
-          kind: 'puppetlabs/ace/reporting_failed',
-          details: 'upstream api errors go here'
-        } }.to_json]
+      begin
+        @plugins.with_synced_libdir(environment, certname) do
+          ACE::TransportApp.init_puppet_target(certname, body['target']['remote-transport'], body['target'])
+          configurer = ACE::Configurer.new(body['compiler']['transaction_uuid'], body['compiler']['job_id'])
+          configurer.run(transport_name: certname,
+                         environment: environment,
+                         network_device: true,
+                         pluginsync: false,
+                         trusted_facts: ACE::TransportApp.trusted_facts(certname))
+        end
+      rescue ACE::Error => e
+        process_error = {  _error: e.to_h }
+        return [400, process_error.to_json]
+      rescue StandardError => e
+        process_error = { _error: ACE::Error.new(e.message,
+                                                 'puppetlabs/ace/processing_exception',
+                                                 class: e.class, backtrace: e.backtrace) }
+        return [400, process_error.to_json]
       else
-        [200, '{}']
+        [200, {
+          certname: certname,
+          status: 'success',
+          result: {
```

Comment:
Any need to return the empty result hash? I guess returning a 200, then the certname, transaction_uuid, job_id in the body would be most ideal as this could link requests to responses if the console team ever requires that granularity? Or in the result hash, have the certname, transation_uuid, and job_id - the `status: 'success'` would be implied by the 200 status code imo

Reply:
I wanted to keep the responses in the same sort of structure, hence an empty `result`. This could in future as you suggest hold info we want to return. I switched out `node` for `certname` and left `status` untouched.

Reply:
But the response to me does not seem to be a similar structure with it having `certname`, `status` at a top level. For the error conditions it is `{ _error: { ... } }` so too keep the response structures of this endpoint similar something like this would seem more appropriate:

```
[200, { _success: {
  certname: certname,
  transaction_uuid: transaction_uuid,
  job_id: job_id
} }.to_json ]
```

Reply:
This is the response from a successful `/run_task`

```
{
    "node": "y7h5ncyip5gcjd8.delivery.puppetlabs.net",
    "target": "y7h5ncyip5gcjd8.delivery.puppetlabs.net",
    "type": null,
    "object": null,
    "status": "success",
    "result": {
        "apikey": "LUFRPT14MW5xOEo1R09KVlBZNnpnemh0VHRBOWl6TGM9bXcwM3JHUGVhRlNiY0dCR0srNERUQT09"
    }
}
```
this is the response from a failed `/run_task`

```
{
    "node": "12345.delivery.puppetlabs.net",
    "target": "12345.delivery.puppetlabs.net",
    "type": null,
    "object": null,
    "status": "failure",
    "result": {
        "_error": {
            "msg": "Failed to open TCP connection to 12345.delivery.puppetlabs.net:443 (getaddrinfo: nodename nor servname provided, or not known)",
            "kind": "puppetlabs-panos/unknown",
            "details": {
                "class": "SocketError"
            }
        }
    }
}
```
This is the response of a successful `/execute_catalog`
```
{
    "certname": "livupbi6o63eyg3.delivery.puppetlabs.net",
    "status": "success",
    "result": {}
}
```
This is the response of a failed `/execute_catalog`
```
{
    "_error": {
        "kind": "puppetlabs/ace/request_exception",
        "msg": "785: unexpected token at '{\n  \"target\":{\n    \"remote-transport\":\"panos\",\n    \"host\":\"livupbi6o63eyg3.delivery.puppetlabs.net\",\n    \"user\":\"admin\"\n    \"password\":\"admin\"\n  },\n  \"compiler\":{\n    \"certname\":\"livupbi6o63eyg3.delivery.puppetlabs.net\",\n    \"environment\":\"production\",\n    \"transaction_uuid\":\"981687ce-520e-11e9-8647-d663bd873d93\",\n    \"job_id\":\"1\"\n  }\n}'",
        "details": {
            "class": "JSON::ParserError",
            "backtrace": [
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/json-2.2.0/lib/json/common.rb:156:in `parse'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/json-2.2.0/lib/json/common.rb:156:in `parse'",
                "/Users/david.armstrong/projects/ace/lib/ace/transport_app.rb:198:in `block in <class:TransportApp>'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1635:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1635:in `block in compile!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:992:in `block (3 levels) in route!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1011:in `route_eval'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:992:in `block (2 levels) in route!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1040:in `block in process_route'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1038:in `catch'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1038:in `process_route'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:990:in `block in route!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:989:in `each'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:989:in `route!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1097:in `block in dispatch!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1076:in `block in invoke'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1076:in `catch'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1076:in `invoke'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1094:in `dispatch!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:924:in `block in call!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1076:in `block in invoke'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1076:in `catch'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1076:in `invoke'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:924:in `call!'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:913:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-protection-2.0.5/lib/rack/protection/xss_header.rb:18:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-protection-2.0.5/lib/rack/protection/path_traversal.rb:16:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-protection-2.0.5/lib/rack/protection/json_csrf.rb:26:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-protection-2.0.5/lib/rack/protection/base.rb:50:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-protection-2.0.5/lib/rack/protection/base.rb:50:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-protection-2.0.5/lib/rack/protection/frame_options.rb:31:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-2.0.7/lib/rack/null_logger.rb:9:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/rack-2.0.7/lib/rack/head.rb:12:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/show_exceptions.rb:22:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:194:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/sinatra-2.0.5/lib/sinatra/base.rb:1957:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/puma-3.12.1/lib/puma/configuration.rb:227:in `call'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/puma-3.12.1/lib/puma/server.rb:660:in `handle_request'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/puma-3.12.1/lib/puma/server.rb:474:in `process_client'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/puma-3.12.1/lib/puma/server.rb:334:in `block in run'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/puma-3.12.1/lib/puma/thread_pool.rb:135:in `block in spawn_thread'",
                "/Users/david.armstrong/projects/ace/.bundle/ruby/2.5.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'"
            ]
        }
    }
}
```

So it looks like the issue is actually how the _error is being returned in comparison